File: src/web/Controllers/OwnershipController.cs
using System;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using core.fs.Adapters.Storage;
using core.Shared;
using web.Utils;

#nullable enable

namespace web.Controllers;

public record SearchEntitiesRequest(string Name);
public record CreateEntityRequest(string Name, string EntityType, string? Cik);
public record CreateRoleRequest(Guid EntityId, string CompanyTicker, string CompanyCik, string RelationshipType, string? Title);
public record CreateOwnershipEventRequest(
    Guid EntityId,
    string CompanyTicker,
    string CompanyCik,
    Guid? FilingId,
    string EventType,
    string? TransactionType,
    long? SharesBefore,
    long? SharesTransacted,
    long SharesAfter,
    decimal? PercentOfClass,
    decimal? PricePerShare,
    decimal? TotalValue,
    string TransactionDate,
    string FilingDate,
    bool IsDirect,
    string? OwnershipNature
);

[ApiController]
[Authorize]
[Route("api/[controller]")]
public class OwnershipController(IOwnershipStorage storage) : ControllerBase
{
    [HttpGet("ticker/{ticker}")]
    public async Task<ActionResult> GetOwnershipByTicker([FromRoute] string ticker)
    {
        var t = new Ticker(ticker);
        var summary = await storage.GetOwnershipSummary(t);
        return Ok(summary);
    }

    [HttpGet("ticker/{ticker}/timeline")]
    public async Task<ActionResult> GetOwnershipTimeline([FromRoute] string ticker, [FromQuery] int days = 365)
    {
        var t = new Ticker(ticker);
        var timeline = await storage.GetOwnershipTimeline(t, days);
        return Ok(timeline);
    }

    [HttpGet("ticker/{ticker}/events")]
    public async Task<ActionResult> GetEventsByTicker([FromRoute] string ticker)
    {
        var t = new Ticker(ticker);
        var events = await storage.GetEventsByCompany(t);
        return Ok(events);
    }

    [HttpGet("entity/{entityId}")]
    public async Task<ActionResult> GetEntity([FromRoute] Guid entityId)
    {
        var entity = await storage.FindEntityByCik(entityId.ToString());
        if (Microsoft.FSharp.Core.OptionModule.IsNone(entity))
        {
            return NotFound();
        }
        return Ok(Microsoft.FSharp.Core.OptionModule.GetValue(entity));
    }

    [HttpGet("entity/{entityId}/roles")]
    public async Task<ActionResult> GetEntityRoles([FromRoute] Guid entityId)
    {
        var roles = await storage.GetRolesByEntity(entityId);
        return Ok(roles);
    }

    [HttpGet("entity/{entityId}/events")]
    public async Task<ActionResult> GetEventsByEntity([FromRoute] Guid entityId)
    {
        var events = await storage.GetEventsByEntity(entityId);
        return Ok(events);
    }

    [HttpGet("entities/search")]
    public async Task<ActionResult> SearchEntities([FromQuery] string name)
    {
        if (string.IsNullOrWhiteSpace(name))
        {
            return BadRequest("Name parameter is required");
        }

        var entities = await storage.FindEntitiesByName(name);
        return Ok(entities);
    }

    [HttpPost("entity")]
    public async Task<ActionResult> CreateEntity([FromBody] CreateEntityRequest request)
    {
        var cikOption = string.IsNullOrWhiteSpace(request.Cik) 
            ? Microsoft.FSharp.Core.FSharpOption<string>.None 
            : Microsoft.FSharp.Core.FSharpOption<string>.Some(request.Cik);

        var entity = new OwnershipEntity
        {
            Id = Guid.NewGuid(),
            Name = request.Name,
            EntityType = request.EntityType,
            Cik = cikOption,
            FirstSeen = DateTimeOffset.UtcNow,
            LastSeen = DateTimeOffset.UtcNow,
            CreatedAt = DateTimeOffset.UtcNow
        };

        var entityId = await storage.SaveEntity(entity);
        return Ok(new { id = entityId });
    }

    [HttpPost("role")]
    public async Task<ActionResult> CreateRole([FromBody] CreateRoleRequest request)
    {
        var titleOption = string.IsNullOrWhiteSpace(request.Title) 
            ? Microsoft.FSharp.Core.FSharpOption<string>.None 
            : Microsoft.FSharp.Core.FSharpOption<string>.Some(request.Title);

        var role = new OwnershipEntityCompanyRole
        {
            Id = Guid.NewGuid(),
            EntityId = request.EntityId,
            CompanyTicker = request.CompanyTicker,
            CompanyCik = request.CompanyCik,
            RelationshipType = request.RelationshipType,
            Title = titleOption,
            IsActive = true,
            FirstSeen = DateTimeOffset.UtcNow,
            LastSeen = DateTimeOffset.UtcNow
        };

        var roleId = await storage.SaveRole(role);
        return Ok(new { id = roleId });
    }

    [HttpPost("event")]
    public async Task<ActionResult> CreateOwnershipEvent([FromBody] CreateOwnershipEventRequest request)
    {
        var filingIdOption = request.FilingId.HasValue 
            ? Microsoft.FSharp.Core.FSharpOption<Guid>.Some(request.FilingId.Value) 
            : Microsoft.FSharp.Core.FSharpOption<Guid>.None;

        var transactionTypeOption = string.IsNullOrWhiteSpace(request.TransactionType) 
            ? Microsoft.FSharp.Core.FSharpOption<string>.None 
            : Microsoft.FSharp.Core.FSharpOption<string>.Some(request.TransactionType);

        var sharesBeforeOption = request.SharesBefore.HasValue 
            ? Microsoft.FSharp.Core.FSharpOption<long>.Some(request.SharesBefore.Value) 
            : Microsoft.FSharp.Core.FSharpOption<long>.None;

        var sharesTransactedOption = request.SharesTransacted.HasValue 
            ? Microsoft.FSharp.Core.FSharpOption<long>.Some(request.SharesTransacted.Value) 
            : Microsoft.FSharp.Core.FSharpOption<long>.None;

        var percentOfClassOption = request.PercentOfClass.HasValue 
            ? Microsoft.FSharp.Core.FSharpOption<decimal>.Some(request.PercentOfClass.Value) 
            : Microsoft.FSharp.Core.FSharpOption<decimal>.None;

        var pricePerShareOption = request.PricePerShare.HasValue 
            ? Microsoft.FSharp.Core.FSharpOption<decimal>.Some(request.PricePerShare.Value) 
            : Microsoft.FSharp.Core.FSharpOption<decimal>.None;

        var totalValueOption = request.TotalValue.HasValue 
            ? Microsoft.FSharp.Core.FSharpOption<decimal>.Some(request.TotalValue.Value) 
            : Microsoft.FSharp.Core.FSharpOption<decimal>.None;

        var ownershipNatureOption = string.IsNullOrWhiteSpace(request.OwnershipNature) 
            ? Microsoft.FSharp.Core.FSharpOption<string>.None 
            : Microsoft.FSharp.Core.FSharpOption<string>.Some(request.OwnershipNature);

        var ownershipEvent = new OwnershipEvent
        {
            Id = Guid.NewGuid(),
            EntityId = request.EntityId,
            CompanyTicker = request.CompanyTicker,
            CompanyCik = request.CompanyCik,
            FilingId = filingIdOption,
            EventType = request.EventType,
            TransactionType = transactionTypeOption,
            SharesBefore = sharesBeforeOption,
            SharesTransacted = sharesTransactedOption,
            SharesAfter = request.SharesAfter,
            PercentOfClass = percentOfClassOption,
            PricePerShare = pricePerShareOption,
            TotalValue = totalValueOption,
            TransactionDate = request.TransactionDate,
            FilingDate = request.FilingDate,
            IsDirect = request.IsDirect,
            OwnershipNature = ownershipNatureOption,
            CreatedAt = DateTimeOffset.UtcNow
        };

        var eventId = await storage.SaveEvent(ownershipEvent);
        return Ok(new { id = eventId });
    }

    [HttpGet("ticker/{ticker}/roles")]
    public async Task<ActionResult> GetRolesByTicker([FromRoute] string ticker)
    {
        var t = new Ticker(ticker);
        var roles = await storage.GetRolesByCompany(t);
        return Ok(roles);
    }
}

File: src/web.fs/Controllers/OwnershipController.fs
namespace web.Controllers

open System
open Microsoft.AspNetCore.Mvc
open Microsoft.AspNetCore.Authorization
open core.fs.Adapters.Storage
open core.Shared

[<CLIMutable>]
type SearchEntitiesRequest = {
    Name: string
}

[<CLIMutable>]
type BatchEntityIdsRequest = {
    Ids: Guid array
}

[<CLIMutable>]
type CreateEntityRequest = {
    Name: string
    EntityType: string
    Cik: string option
}

[<CLIMutable>]
type CreateRoleRequest = {
    EntityId: Guid
    CompanyTicker: string
    CompanyCik: string
    RelationshipType: string
    Title: string option
}

[<CLIMutable>]
type CreateOwnershipEventRequest = {
    EntityId: Guid
    CompanyTicker: string
    CompanyCik: string
    FilingId: Guid option
    EventType: string
    TransactionType: string option
    SharesBefore: int64 option
    SharesTransacted: int64 option
    SharesAfter: int64
    PercentOfClass: decimal option
    PricePerShare: decimal option
    TotalValue: decimal option
    TransactionDate: string
    FilingDate: string
    IsDirect: bool
    OwnershipNature: string option
}

[<ApiController>]
[<Authorize>]
[<Route("api/[controller]")>]
type OwnershipController(storage: IOwnershipStorage) =
    inherit ControllerBase()

    [<HttpGet("ticker/{ticker}")>]
    member this.GetOwnershipByTicker([<FromRoute>] ticker: string) = task {
        let t = Ticker ticker
        let! summary = storage.GetOwnershipSummary(t)
        return this.Ok summary
    }

    [<HttpGet("ticker/{ticker}/timeline")>]
    member this.GetOwnershipTimeline([<FromRoute>] ticker: string, [<FromQuery>] days: Nullable<int>) = task {
        let t = Ticker ticker
        let d = if days.HasValue then days.Value else 1825
        let! timeline = storage.GetOwnershipTimeline t d
        return this.Ok timeline
    }

    [<HttpGet("ticker/{ticker}/events")>]
    member this.GetEventsByTicker([<FromRoute>] ticker: string) = task {
        let t = Ticker ticker
        let! events = storage.GetEventsByCompany(t)
        return this.Ok events
    }

    [<HttpGet("entity/{entityId}")>]
    member this.GetEntity([<FromRoute>] entityId: Guid) = task {
        let! entity = storage.GetEntityById entityId
        match entity with
        | Some e -> return this.Ok(e) :> ActionResult
        | None -> return this.NotFound() :> ActionResult
    }

    [<HttpGet("entity/{entityId}/roles")>]
    member this.GetEntityRoles([<FromRoute>] entityId: Guid) = task {
        let! roles = storage.GetRolesByEntity entityId
        return this.Ok(roles)
    }

    [<HttpGet("entity/{entityId}/events")>]
    member this.GetEventsByEntity([<FromRoute>] entityId: Guid) = task {
        let! events = storage.GetEventsByEntity entityId
        return this.Ok(events)
    }

    [<HttpPost("entities/batch")>]
    member this.GetEntitiesBatch([<FromBody>] request: BatchEntityIdsRequest) = task {
        if isNull (box request) || request.Ids = null || request.Ids.Length = 0 then
            return this.Ok(Array.empty) :> ActionResult
        else
            let! entities = storage.GetEntitiesByIds(request.Ids)
            return this.Ok entities :> ActionResult
    }

    [<HttpGet("entities/search")>]
    member this.SearchEntities([<FromQuery>] name: string) = task {
        if String.IsNullOrWhiteSpace name then
            return this.BadRequest "Name parameter is required" :> ActionResult
        else
            let! entities = storage.FindEntitiesByName name
            return this.Ok entities :> ActionResult
    }

    [<HttpGet("timelines/recent")>]
    member this.GetRecentTimelines([<FromQuery>] limit: Nullable<int>) = task {
        let l = if limit.HasValue then limit.Value else 50
        let! timelines = storage.GetRecentTimelines l
        return this.Ok timelines
    }

    [<HttpPost("entity")>]
    member this.CreateEntity([<FromBody>] request: CreateEntityRequest) = task {
        let entity = {
            Id = Guid.NewGuid()
            Name = request.Name
            EntityType = request.EntityType
            Cik = request.Cik
            FirstSeen = DateTimeOffset.UtcNow
            LastSeen = DateTimeOffset.UtcNow
            CreatedAt = DateTimeOffset.UtcNow
        }
        let! entityId = storage.SaveEntity entity
        return this.Ok {| id = entityId |}
    }

    [<HttpPost("role")>]
    member this.CreateRole([<FromBody>] request: CreateRoleRequest) = task {
        let role = {
            Id = Guid.NewGuid()
            EntityId = request.EntityId
            CompanyTicker = request.CompanyTicker
            CompanyCik = request.CompanyCik
            RelationshipType = request.RelationshipType
            Title = request.Title
            IsActive = true
            FirstSeen = DateTimeOffset.UtcNow
            LastSeen = DateTimeOffset.UtcNow
        }
        let! roleId = storage.SaveRole role
        return this.Ok {| id = roleId |}
    }

    [<HttpPost("event")>]
    member this.CreateOwnershipEvent([<FromBody>] request: CreateOwnershipEventRequest) = task {
        let ownershipEvent = {
            Id = Guid.NewGuid()
            EntityId = request.EntityId
            CompanyTicker = request.CompanyTicker
            CompanyCik = request.CompanyCik
            FilingId = request.FilingId
            EventType = request.EventType
            TransactionType = request.TransactionType
            SharesBefore = request.SharesBefore
            SharesTransacted = request.SharesTransacted
            SharesAfter = request.SharesAfter
            PercentOfClass = request.PercentOfClass
            PricePerShare = request.PricePerShare
            TotalValue = request.TotalValue
            TransactionDate = request.TransactionDate
            FilingDate = request.FilingDate
            IsDirect = request.IsDirect
            OwnershipNature = request.OwnershipNature
            CreatedAt = DateTimeOffset.UtcNow
        }
        let! eventId = storage.SaveEvent ownershipEvent
        return this.Ok {| id = eventId |}
    }

    [<HttpGet("ticker/{ticker}/roles")>]
    member this.GetRolesByTicker([<FromRoute>] ticker: string) = task {
        let t = Ticker ticker
        let! roles = storage.GetRolesByCompany t
        return this.Ok roles
    }
